firebase/src/core/data/gateways/firestore-gateway.ts

```diff
@@ -0,0 +1,235 @@
+import * as firebase from "firebase-admin";
+import FirebaseFirestoreError from "@faults/errors/firebase-firestore-error";
+
+type Firestore = firebase.firestore.Firestore;
+type Transaction = firebase.firestore.Transaction;
+type CollectionReferece = firebase.firestore.CollectionReference;
+type CollectionGroup = firebase.firestore.CollectionGroup;
+type Query = firebase.firestore.Query;
+type DocumentSnapshot = firebase.firestore.DocumentSnapshot;
+type DocumentData = firebase.firestore.DocumentData;
+
+type QueryComparison = FirebaseFirestore.WhereFilterOp;
+
+/**
+ * Represent a `Firestore` where filter, used in read operations.
+ *
+ * @param field Object's field name to be compared
+ * @param comparison Filter operation
+ * @param value Expected field value
+ */
+export interface QueryFilter {
+  field: string;
+  comparison: QueryComparison;
+  value: unknown;
+}
+
+/** Maximum operations for a single transaction or batch of writes, inclusive. */
+export const firestoreOperationsLimit = 499;
+
+/** Wraps {@link Firestore} operations, allowing transactions and typed collections. */
+export class FirestoreGateway<Path extends string> {
+  readonly #firestore: Firestore;
+
+  constructor(firestore: Firestore) {
+    this.#firestore = firestore;
+  }
+
+  /** Current transaction, if there is an ongoing {@link FirestoreGateway.runTransaction}. */
+  #transaction: Transaction | null = null;
```

Comment:
I'm not sure how the use-cases will be instantiated, but using a Transaction here imply that only a single instance of this gateway will be used across the function container. If that it's expected, let's move on.

Reply:
Given my past experiences, I've never instantiated multiple gateways. Encapsulating the transaction lifecycle gives much more freedom and less verbosity to this implementation, while providing the same functionality if using an external "controller" to this lifecycle.

Reply:
I'm pointing some potential issues that we already faced when using multiple transactions. Consider the example of running multiple transactions in parallel:

```typescript

await Promise.all([
  firestore.runTransaction(() => {
    // Do some stuff here  
  }),
  firestore.runTransaction(() => {
    // Do other stuff here  
  }),
])

```

The transactions would overwrite each other, since the same gateway is being used. This is the only use-case limitation that I faced when using a transaction inside the gateway.

Reply:
I think this is a problem with the implementation usage itself, and not the logic.

Maybe we will eventually face the same problems, but as it stands right now, there is no need to add more complexity unless we know we will need it.

Reply:
Agreed, just wanted to state because we faced in the past. If we start doing heavy jobs and parallel transactions, we'll need to rethink how to handle with this transaction (and probably use providers to ensure only one transaction is used). But I agree about keeping it simple for now, but we must keep this specific problem in mind.

Reply:
Agreed, just wanted to state because we faced in the past. If we start doing heavy jobs and parallel transactions, we'll need to rethink how to handle with this transaction (and probably use providers to ensure only one transaction is used). But I agree about keeping it simple for now, but we must keep this specific problem in mind.

Reply:
I also think that there are better solutions to handle the transaction management by not exposing it to a parent, but that's a completely different discussion